README.md
# Windrose Chart Component

![Node.js Package](https://github.com/eunchurn/windrose-chart/workflows/Node.js%20Package/badge.svg) ![Unit Test](https://github.com/eunchurn/windrose-chart/workflows/Unit%20Test/badge.svg?branch=master) ![Deploy Storybook](https://github.com/eunchurn/windrose-chart/workflows/Deploy%20Storybook/badge.svg) [![codecov](https://codecov.io/gh/eunchurn/windrose-chart/branch/master/graph/badge.svg)](https://codecov.io/gh/eunchurn/windrose-chart) [![npm version](https://badge.fury.io/js/%40eunchurn%2Fwindrose-chart.svg)](https://badge.fury.io/js/%40eunchurn%2Fwindrose-chart) ![CodeQL](https://github.com/eunchurn/windrose-chart/workflows/CodeQL/badge.svg) [![CLA assistant](https://cla-assistant.io/readme/badge/eunchurn/windrose-chart)](https://cla-assistant.io/eunchurn/windrose-chart) [![pages-build-deployment](https://github.com/eunchurn/windrose-chart/actions/workflows/pages/pages-build-deployment/badge.svg)](https://github.com/eunchurn/windrose-chart/actions/workflows/pages/pages-build-deployment)


![chart](doc/2020-06-09%2002.59.40.png)

## Features

- INPUT: Professionals respond to survey of how much they use a K-12 core competancy in each subject
- OUTPUT: Interdisciplinarity Badge (ideally with mouseover tooltips). Wind-rose type graphic, displaying survey response magnitudes for each subject area core competency

### Demo

<https://eunchurn.github.io/windrose-chart>

## Usage

### Install

```bash
npm install @eunchurn/windrose-chart
```

or

```bash
yarn install @eunchurn/windrose-chart
```

### React Component (defaultProps)

- Refer `DefaultProps.tsx` data format

```tsx
import Chart from "@eunchurn/windrose-chart";

export default function WindRoseChart() {
  return (
    <Chart />
  )
}
```

- With `props`

### React Component (Props)

```tsx
import Chart from "@eunchurn/windrose-chart"

export default function WindRoseChart() {
  return (
    <Chart 
      data={data} 
      width={800} 
      height={500} 
      dataMax={5} 
      columns={columns} 
      angles={angles}
      dataKeys={["survey"]}
    />
  )
};
```

### Data Type

```typescript
export interface DataType {
  [key: string]: number | null;
}

export interface StackKey {
  key: string;
}
export interface State {
  width: number;
  height: number;
}

export interface PropType extends State {
  /**
   * Professionals respond to survey of how much they use a K-12 core competancy in each subject
   */
  data: DataType[];
  /**
   * Subjects
   */
  columns: string[];
  /**
   * Subjects colors
   */
  columnsColor: string[];
  /**
   * All core competency
   */
  angles: string[];
  /**
   * Max score
   */
  dataMax: number;
  /**
   * Target data keys
   */
  dataKeys: string[];
  /**
   * Mouse over Path color
   */
  mouseOverColor: string;
  /**
   * Mouse over competency text color
   */
  mouseOverTitleColor: string;
  /**
   * Mouseover survey score text color
   */
  mouseOverSurveyColor: string;
}

```

## Default data

```typescript
const data = [
  {
    subject: "Language Arts",
    coreCompetency: "Reading/Verbal Comprehension",
    survey: 1.2,
    color: "#8e44ad",
  },
  {
    subject: "Language Arts",
    coreCompetency: "Writing",
    survey: 2.3,
    color: "#7e44ad",
  },
  {
    subject: "Language Arts",
    coreCompetency: "Textual Analysis",
    survey: 2.4,
    color: "#6e44ad",
  },
  {
    subject: "Language Arts",
    coreCompetency: "Literary Knowledge",
    survey: 3.1,
    color: "#5e44ad",
  },
  {
    subject: "Language Arts",
    coreCompetency: "Foreign Language",
    survey: 1.7,
    color: "#4e44ad",
  },
  {
    subject: "Math",
    coreCompetency: "Arithmetic/ Algebra",
    survey: 4.2,
    color: "#4e54ad",
  },
  {
    subject: "Math",
    coreCompetency: "Geometry/Trig/Spatial Orientation",
    survey: 5.6,
    color: "#4e64ad",
  },
  {
    subject: "Math",
    coreCompetency: "Probability/Statistics/Modeling",
    survey: 1.6,
    color: "#4e74ad",
  },
  {
    subject: "Math",
    coreCompetency: "Calculus/Other advanced math",
    survey: 2.3,
    color: "#4e84ad",
  },
  {
    subject: "Math",
    coreCompetency: "Graphical Literacy",
    survey: 5.6,
    color: "#4e94ad",
  },
  {
    subject: "Science",
    coreCompetency:
      "Application of Scientific method or deductive reasoning",
    survey: 5.8,
    color: "#4e949d",
  },
  {
    subject: "Science",
    coreCompetency: "Experimental/Prototype design and revision",
    survey: 5.2,
    color: "#4e948d",
  },
  {
    subject: "Science",
    coreCompetency:
      "Synthesis or inference based on multiple lines of evidence",
    survey: 5.1,
    color: "#4e947d",
  },
  {
    subject: "Science",
    coreCompetency:
      "Disciplinary Knowledge (Physics, biology, chemistry, medicine, engineering, etc.)",
    survey: 5.7,
    color: "#4e946d",
  },
  {
    subject: "Science",
    coreCompetency: "Data management/visualization",
    survey: 4.2,
    color: "#4e945d",
  },
  {
    subject: "Social Studies",
    coreCompetency:
      "Constructing questions, gathering and synthesizing sources of cultural, geographical, and historical knowledge",
    survey: 1.3,
    color: "#3e945d",
  },
  {
    subject: "Social Studies",
    coreCompetency: "Understanding of government, law, & politics",
    survey: 4.9,
    color: "#2e945d",
  },
  {
    subject: "Social Studies",
    coreCompetency:
      "World history and geography: temporal and spatial views of the world, human-environmental interactions, (changing) spatial patterns and movement",
    survey: 1.2,
    color: "#1e945d",
  },
  {
    subject: "Social Studies",
    coreCompetency:
      "Higher-level analysis: Evaluating sources of cultural and historical knowledge; developing claims from evidence",
    survey: 2.3,
    color: "#10945d",
  },
  {
    subject: "Social Studies",
    coreCompetency:
      "Communicating and critiquing historical, political, economic, or cultural viewpoints",
    survey: 1.1,
    color: "#00945d",
  },
  {
    subject: "21st Century Skills",
    coreCompetency: "Collaboration",
    survey: 5.6,
    color: "#00845d",
  },
  {
    subject: "21st Century Skills",
    coreCompetency: "Communication",
    survey: 5.1,
    color: "#00745d",
  },
  {
    subject: "21st Century Skills",
    coreCompetency: "Creativity",
    survey: 5.3,
    color: "#00645d",
  },
  {
    subject: "21st Century Skills",
    coreCompetency: "Critical Thinking",
    survey: 4.5,
    color: "#00545d",
  },
  {
    subject: "21st Century Skills",
    coreCompetency:
      "Tech Savvy (coding, touch-typing, troubleshooting abilities, software competency)",
    survey: 5.9,
    color: "#00445d",
  },
  {
    subject: "Arts, Crafts & Labor",
    coreCompetency: "Physical Effort",
    survey: 1.9,
    color: "#FFFF00",
  },
  {
    subject: "Arts, Crafts & Labor",
    coreCompetency: "Mental Effort",
    survey: 5.8,
    color: "#FFFF10",
  },
  {
    subject: "Arts, Crafts & Labor",
    coreCompetency: "Emotional Effort",
    survey: 2.1,
    color: "#FFFF20",
  },
  {
    subject: "Arts, Crafts & Labor",
    coreCompetency:
      "Specialized art or craft knowledge (sketching, painting, carpentry, hair cutting, welding, etc)",
    survey: 1.9,
    color: "#FFFF30",
  },
  {
    subject: "Arts, Crafts & Labor",
    coreCompetency: "Appreciation of or critique of art/craft/design",
    survey: 1.7,
    color: "#FFFF40",
  },
];

const angles = data.map((d) => d.coreCompetency);
const maxData = data.reduce((pre, cur) =>
  pre.survey > cur.survey ? pre : cur
);
```

### Default Props

```typescript
export const DefaultProps = {
  width: 800, // Chart Width (px)
  height: 500,  // Chart Height (px)
  dataMax: maxData.survey, // Max data value (5)
  data, // Data
  columns: [
    "Language Arts",
    "Math",
    "Science",
    "Social Studies",
    "21st Century Skills",
    "Arts, Crafts & Labor",
  ], // Columns of group (array string)
    columnsColor: [
    "#8e44ad",
    "#c0392b",
    "#27ae60",
    "#2c3e50",
    "#0984e3",
    "#e84393",
  ],
  angles, // All pie chart angles (array string)
  dataKeys: ["survey"], // target keys of `data` / legnth = 1, no stack data in this time
  mouseOverColor: "#1abc9c",
  mouseOverTitleColor: "#e67e22",
  mouseOverSurveyColor: "#e74c3c",
};
```

## Development

### Install

- Clone this repo
- install packages

```bash
npm i
```

### Storybook

```bash
npm run storybook
```

## Pull request

Feel free to create a pull request, <https://github.com/eunchurn/windrose-chart/pulls>

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.2](https://github.com/eunchurn/windrose-chart/compare/v0.2.1...v0.2.2) (2022-05-02)


### Bug Fixes

* **deps:** update ([4a70440](https://github.com/eunchurn/windrose-chart/commit/4a70440930a41d8cd8fddf565a2e04b2bc2568f0))
* **deps:** update react monorepo to v17 ([ed32c3c](https://github.com/eunchurn/windrose-chart/commit/ed32c3cdf72da7b2df24ceb0a6d6a4cb167891c3))
* **minimist:** vulnerability fix ([919bc9a](https://github.com/eunchurn/windrose-chart/commit/919bc9a50d7b47ea77c41046fc5feba4424fe83b))
* **script:** ci ([ab7aa0e](https://github.com/eunchurn/windrose-chart/commit/ab7aa0ea25aed5e64b0f7b1d1c74d5d0190504bf))

### [0.2.1](https://github.com/eunchurn/windrose-chart/compare/v0.2.0...v0.2.1) (2022-02-27)


### Bug Fixes

* **script:** npmpublish ([eefe8f7](https://github.com/eunchurn/windrose-chart/commit/eefe8f7300b9152420d70e484ace3bab36e4902e))

## [0.2.0](https://github.com/eunchurn/windrose-chart/compare/v0.1.1...v0.2.0) (2022-02-27)


### Features

* **npm to yarn:** for vulnerability ([f49b0bc](https://github.com/eunchurn/windrose-chart/commit/f49b0bc00e03fb4eb782f7b57b0e89954e38dd7b))
* **npm to yarn:** for vulnerability ([48e4211](https://github.com/eunchurn/windrose-chart/commit/48e421169a8068c6b38b984f4a488acddabba6c6))
* **upgrade:** storybook ([8b80695](https://github.com/eunchurn/windrose-chart/commit/8b80695a9b2a62884e16a806c952aadd21c7602b))


### Bug Fixes

* **deps:** update all ([963bb93](https://github.com/eunchurn/windrose-chart/commit/963bb935d95420c4a160cd04cd6b620088b0f8b7))
* **deps:** update dependency d3 to ^6.5.0 ([3061f69](https://github.com/eunchurn/windrose-chart/commit/3061f69ec7b2e5778d7ee744889d3244e4931205))
* **jest:** config ([b55be7b](https://github.com/eunchurn/windrose-chart/commit/b55be7bb64e0b34410caee01a0268f52bedb0598))
* **script:** build-story ([583ffbb](https://github.com/eunchurn/windrose-chart/commit/583ffbb692b6a8b04baefe635633c331459ae4d1))
* **script:** node version ([458843f](https://github.com/eunchurn/windrose-chart/commit/458843f6b3c8e3332c41a93b6589f07eaf1f9b1b))
* **snapshot:** update ([08a38b4](https://github.com/eunchurn/windrose-chart/commit/08a38b41c8d852e81c2e7e4147dab3b9294a951a))
* **snapshot:** update ([464cbe0](https://github.com/eunchurn/windrose-chart/commit/464cbe0a9138b32c68426668d374728ea1e573b6))
* **style:** outline none ([7a9a70f](https://github.com/eunchurn/windrose-chart/commit/7a9a70f879516a49e1c5f35fd787dab01f7f9b8e))
* upgrade d3 from 6.3.0 to 6.3.1 ([5139647](https://github.com/eunchurn/windrose-chart/commit/5139647f030698fa6372db04d7aa3a0dfd4a8860))
* upgrade react-tooltip from 4.2.10 to 4.2.21 ([fb4c33c](https://github.com/eunchurn/windrose-chart/commit/fb4c33c466cbd92e115220c0e570b945e046f6d3))

### [0.1.1](https://github.com/eunchurn/windrose-chart/compare/v0.1.0...v0.1.1) (2020-12-25)

## [0.1.0](https://github.com/eunchurn/windrose-chart/compare/v0.0.2...v0.1.0) (2020-12-25)


### Bug Fixes

* classname on react-tooltip jest mock ([5cad13e](https://github.com/eunchurn/windrose-chart/commit/5cad13ef52952c8d2054b7f967034cc409c54f5e))
* package.json, package-lock.json & .snyk to reduce vulnerabilities ([8560ea2](https://github.com/eunchurn/windrose-chart/commit/8560ea285008de3e36114b40503a3f49ca0743e8))
* remove bump ([fe45aca](https://github.com/eunchurn/windrose-chart/commit/fe45aca8221f9017bec021f50a3ebe2131b7f325))
* update snapshot ([7085e7e](https://github.com/eunchurn/windrose-chart/commit/7085e7e67a369d20154d5ff24ea560264760651e))
* update snapshot ([bfa3dc2](https://github.com/eunchurn/windrose-chart/commit/bfa3dc2a047133ea22ba88403995b8169f4a1955))
* upgrade d3 from 6.2.0 to 6.3.0 ([c1cdf3c](https://github.com/eunchurn/windrose-chart/commit/c1cdf3c377127a29f42728d698a8e4ea4b153b93))
* upgrade snyk from 1.336.0 to 1.337.0 ([08b72b9](https://github.com/eunchurn/windrose-chart/commit/08b72b973bac47e5124482e659839c81c3e720fb))


index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var d3 = require('d3');
var styled = require('styled-components');
var lodash = require('lodash');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var styled__default = /*#__PURE__*/_interopDefaultLegacy(styled);

function _taggedTemplateLiteral(strings, raw) {
  if (!raw) {
    raw = strings.slice(0);
  }

  return Object.freeze(Object.defineProperties(strings, {
    raw: {
      value: Object.freeze(raw)
    }
  }));
}

var data = [{
  subject: "Language Arts",
  coreCompetency: "Reading/Verbal Comprehension",
  survey: 1.2,
  color: "#8e44ad"
}, {
  subject: "Language Arts",
  coreCompetency: "Writing",
  survey: 2.3,
  color: "#7e44ad"
}, {
  subject: "Language Arts",
  coreCompetency: "Textual Analysis",
  survey: 2.4,
  color: "#6e44ad"
}, {
  subject: "Language Arts",
  coreCompetency: "Literary Knowledge",
  survey: 3.1,
  color: "#5e44ad"
}, {
  subject: "Language Arts",
  coreCompetency: "Foreign Language",
  survey: 1.7,
  color: "#4e44ad"
}, {
  subject: "Math",
  coreCompetency: "Arithmetic/ Algebra",
  survey: 4.2,
  color: "#c0392b"
}, {
  subject: "Math",
  coreCompetency: "Geometry/Trig/Spatial Orientation",
  survey: 5.6,
  color: "#c0492b"
}, {
  subject: "Math",
  coreCompetency: "Probability/Statistics/Modeling",
  survey: 1.6,
  color: "#c0592b"
}, {
  subject: "Math",
  coreCompetency: "Calculus/Other advanced math",
  survey: 2.3,
  color: "#c0692b"
}, {
  subject: "Math",
  coreCompetency: "Graphical Literacy",
  survey: 5.6,
  color: "#c0792b"
}, {
  subject: "Science",
  coreCompetency: "Application of Scientific method or deductive reasoning",
  survey: 5.8,
  color: "#27ae60"
}, {
  subject: "Science",
  coreCompetency: "Experimental/Prototype design and revision",
  survey: 5.2,
  color: "#27ae70"
}, {
  subject: "Science",
  coreCompetency: "Synthesis or inference based on multiple lines of evidence",
  survey: 5.1,
  color: "#27ae80"
}, {
  subject: "Science",
  coreCompetency: "Disciplinary Knowledge (Physics, biology, chemistry, medicine, engineering, etc.)",
  survey: 5.7,
  color: "#27ae90"
}, {
  subject: "Science",
  coreCompetency: "Data management/visualization",
  survey: 4.2,
  color: "#27aea0"
}, {
  subject: "Social Studies",
  coreCompetency: "Constructing questions, gathering and synthesizing sources of cultural, geographical, and historical knowledge",
  survey: 1.3,
  color: "#2c3e50"
}, {
  subject: "Social Studies",
  coreCompetency: "Understanding of government, law, & politics",
  survey: 4.9,
  color: "#2c4e50"
}, {
  subject: "Social Studies",
  coreCompetency: "World history and geography: temporal and spatial views of the world, human-environmental interactions, (changing) spatial patterns and movement",
  survey: 1.2,
  color: "#2c5e50"
}, {
  subject: "Social Studies",
  coreCompetency: "Higher-level analysis: Evaluating sources of cultural and historical knowledge; developing claims from evidence",
  survey: 2.3,
  color: "#2c6e50"
}, {
  subject: "Social Studies",
  coreCompetency: "Communicating and critiquing historical, political, economic, or cultural viewpoints",
  survey: 1.1,
  color: "#2c7e50"
}, {
  subject: "21st Century Skills",
  coreCompetency: "Collaboration",
  survey: 5.6,
  color: "#0984e3"
}, {
  subject: "21st Century Skills",
  coreCompetency: "Communication",
  survey: 5.1,
  color: "#0974e3"
}, {
  subject: "21st Century Skills",
  coreCompetency: "Creativity",
  survey: 5.3,
  color: "#0964e3"
}, {
  subject: "21st Century Skills",
  coreCompetency: "Critical Thinking",
  survey: 4.5,
  color: "#0954e3"
}, {
  subject: "21st Century Skills",
  coreCompetency: "Tech Savvy (coding, touch-typing, troubleshooting abilities, software competency)",
  survey: 5.9,
  color: "#0944e3"
}, {
  subject: "Arts, Crafts & Labor",
  coreCompetency: "Physical Effort",
  survey: 1.9,
  color: "#e84393"
}, {
  subject: "Arts, Crafts & Labor",
  coreCompetency: "Mental Effort",
  survey: 5.8,
  color: "#d84393"
}, {
  subject: "Arts, Crafts & Labor",
  coreCompetency: "Emotional Effort",
  survey: 2.1,
  color: "#c84393"
}, {
  subject: "Arts, Crafts & Labor",
  coreCompetency: "Specialized art or craft knowledge (sketching, painting, carpentry, hair cutting, welding, etc)",
  survey: 1.9,
  color: "#b84393"
}, {
  subject: "Arts, Crafts & Labor",
  coreCompetency: "Appreciation of or critique of art/craft/design",
  survey: 1.7,
  color: "#a84393"
}];
var angles = data.map(function (d) {
  return d.coreCompetency;
});
var maxData = data.reduce(function (pre, cur) {
  return pre.survey > cur.survey ? pre : cur;
});
var DefaultProps = {
  width: 600,
  height: 500,
  dataMax: maxData.survey,
  data: data,
  columns: ["Language Arts", "Math", "Science", "Social Studies", "21st Century Skills", "Arts, Crafts & Labor"],
  columnsColor: ["#8e44ad", "#c0392b", "#27ae60", "#2c3e50", "#0984e3", "#e84393"],
  angles: angles,
  dataKeys: ["survey"],
  mouseOverColor: "#1abc9c",
  mouseOverTitleColor: "#e67e22",
  mouseOverSurveyColor: "#e74c3c"
};

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  .axis {\n    position: absolute;\n    top: 0;\n    left: 0;\n    stroke: gray;\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  position: relative;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
function Chart(props) {
  var width = props.width,
      height = props.height,
      data = props.data,
      columns = props.columns,
      z = props.columnsColor,
      angles = props.angles,
      dataMax = props.dataMax,
      dataKeys = props.dataKeys,
      mouseOverColor = props.mouseOverColor,
      mouseOverTitleColor = props.mouseOverTitleColor,
      mouseOverSurveyColor = props.mouseOverSurveyColor;
  var containerRef = React.useRef(null);
  var axisContainerRef = React.useRef(null);
  React.useEffect(function () {
    var svg = d3.select(containerRef.current);
    svg.selectAll("*").remove();
    var margin = {
      top: 80,
      right: 100,
      bottom: 80,
      left: 40
    };
    var innerRadius = 20;
    var chartWidth = width - margin.left - margin.right;
    var chartHeight = height - margin.top - margin.bottom;
    var outerRadius = innerRadius + Math.min(chartWidth, chartHeight) / 2;
    var g = svg.append("g").attr("transform", "translate(".concat(width / 2, ", ").concat(height / 2, ")"));
    var angle = d3.scaleLinear().range([0, 2 * Math.PI]);
    var radius = d3.scaleLinear().range([innerRadius, outerRadius]);
    var x = d3.scaleBand().range([0, 2 * Math.PI]).align(0);
    var xGroup = d3.scaleBand().range([0, 2 * Math.PI]).align(0);
    var y = d3.scaleLinear().range([innerRadius, outerRadius]);
    x.domain(angles.map(function (d) {
      return d;
    }));
    xGroup.domain(columns.map(function (d) {
      return d;
    }));
    y.domain([0, d3.max(data, function (_ref) {
      var total = _ref.total;
      return total;
    }) > dataMax ? d3.max(data, function (_ref2) {
      var total = _ref2.total;
      return total;
    }) : dataMax]);
    angle.domain([0, d3.max(data, function (_, i) {
      return i + 1;
    })]);
    radius.domain([0, d3.max(data, function () {
      return 0;
    })]);
    var angleOffset = -360.0 / data.length / 2.0;
    var stackGen = d3.stack().keys(dataKeys);
    var arcVal = d3.arc().innerRadius(function (d) {
      return Number(y(d[0]));
    }).outerRadius(function (d) {
      return Number(y(d[1]));
    }).startAngle(function (_d, i) {
      return Number(x(angles[i]));
    }).endAngle(function (_d, i) {
      return Number(x(angles[i])) + x.bandwidth();
    }).padAngle(0.0).padRadius(innerRadius);
    var arcParent = g.append("g").selectAll("g").data(stackGen(data)).enter();
    var arc = arcParent.selectAll("path").data(function (d) {
      return d;
    }).enter().append("path");
    arc // @ts-ignore
    .attr("d", arcVal).attr("transform", "rotate(".concat(angleOffset, ")")).attr("fill", function (d) {
      return d.data.color;
    }).attr("class", function (_d, i) {
      return "arc_".concat(i);
    }).on("mouseover", lodash.debounce(function (_event, d) {
      var classname = this.getAttribute("class") || "arc_1";
      var i = classname.split("_")[1];
      this.setAttribute("fill", mouseOverColor);
      this.setAttribute("style", "transition: fill 0.5s; cursor: pointer;");
      var bbox = this.getBBox();
      arcParent.append("text").attr("class", "tooltip_".concat(i)).attr("x", bbox.x + bbox.width / 2).attr("dy", bbox.y + bbox.height / 2).attr("text-anchor", "middle").style("font-size", 12).style("fill", mouseOverTitleColor).style("opacity", 1).style("transition", "opacity 0.2s").attr("font-weight", 600).text("".concat(d.data.coreCompetency));
      arcParent.append("text").attr("class", "tooltip_".concat(i)).attr("x", bbox.x + bbox.width / 2).attr("dy", bbox.y + bbox.height / 2 + 14).attr("text-anchor", "middle").style("font-size", 14).style("fill", mouseOverSurveyColor).style("opacity", 1).style("transition", "opacity 0.2s").attr("font-weight", 800).text("".concat(d.data.survey));
    }, 0)).on("mouseout", lodash.debounce(function (_event, d) {
      var classname = this.getAttribute("class") || "arc_1";
      var i = classname.split("_")[1];
      this.setAttribute("fill", String(d.data.color) || "#ffffff"); // this.setAttribute("style", "opacity: 0");

      d3.selectAll(".tooltip_".concat(i)).remove();
    }, 0));
    var label = g.append("g").selectAll("g").data(columns).enter().append("g").attr("text-anchor", "middle").attr("transform", function (d) {
      if (typeof d === "undefined") return null;
      return "rotate(".concat((Number(xGroup(d)) + xGroup.bandwidth() / 2) * 180 / Math.PI - (90 - angleOffset), ")translate(").concat(outerRadius, ",0)");
    });
    label.append("text").attr("transform", "rotate(90)translate(0,-9)").text(function (_d, i) {
      return columns[i];
    }).style("font-size", 14);
    var yAxis = g.append("g").attr("text-anchor", "middle");
    var yTick = yAxis.selectAll("g").data(y.ticks(5).slice(1)).enter().append("g");
    yTick.append("circle").attr("fill", "none").attr("stroke", "gray").attr("stroke-dasharray", "4,4") // @ts-ignore
    .attr("r", y);
    yTick.append("text") // @ts-ignore
    .attr("y", function (d) {
      return -y(d);
    }).attr("dy", "-0.35em").attr("x", function () {
      return -10;
    }).text(y.tickFormat(5, "s")).style("font-size", 14);
    var legend = g.append("g").selectAll("g").data(columns).enter().append("g").attr("transform", function (_d, i) {
      return "translate(".concat(outerRadius + 0, ",").concat(-outerRadius + 40 + (i - (columns.length - 1) / 2) * 20, ")");
    });
    legend.append("rect").attr("width", 18).attr("height", 18).attr("fill", function (_d, i) {
      return z[i];
    });
    legend.append("text").attr("x", 24).attr("y", 9).attr("dy", "0.35em").text(function (d) {
      return d;
    }).style("font-size", 12);
    g.exit().remove();
  });
  return /*#__PURE__*/React.createElement(AxisContainer, {
    ref: axisContainerRef,
    role: "main"
  }, /*#__PURE__*/React.createElement(Axis, {
    className: "axis",
    width: width,
    height: height,
    ref: containerRef,
    role: "document"
  }));
}
var AxisContainer = styled__default['default'].div(_templateObject());
var Axis = styled__default['default'].svg(_templateObject2());
var data$1 = DefaultProps.data,
    angles$1 = DefaultProps.angles,
    columns = DefaultProps.columns,
    columnsColor = DefaultProps.columnsColor,
    width = DefaultProps.width,
    height = DefaultProps.height,
    dataMax = DefaultProps.dataMax,
    dataKeys = DefaultProps.dataKeys,
    mouseOverColor = DefaultProps.mouseOverColor,
    mouseOverTitleColor = DefaultProps.mouseOverTitleColor,
    mouseOverSurveyColor = DefaultProps.mouseOverSurveyColor;
Chart.defaultProps = {
  data: data$1,
  dataMax: dataMax,
  angles: angles$1,
  columns: columns,
  columnsColor: columnsColor,
  width: width,
  height: height,
  dataKeys: dataKeys,
  mouseOverColor: mouseOverColor,
  mouseOverTitleColor: mouseOverTitleColor,
  mouseOverSurveyColor: mouseOverSurveyColor
};

/* eslint-disable import/no-named-as-default */

exports.Axis = Axis;
exports.AxisContainer = AxisContainer;
exports.Chart = Chart;
exports.default = Chart;
//# sourceMappingURL=index.js.map
